### DISCUSSION TITLE
What does the "Data Frame" format look like and when should I use it?

### DISCUSSION THREAD

**Discussion by svet-b** [grafana/grafana-infinity-datasource#512] [Q&A]
I'm often somewhat stumped by the "Format" drop-down in the source configuration, and in particular by the meaning of the "Data Frame" format.

Firstly - and maybe a silly question - does this relate to the format of the data coming from the source (API endpoint), or the data format after processing (e.g. via UQL)?

More specifically, what would a Data Frame look like, for example if coming from a JSON API? I'm asking since I have quite a lot of control over the data format from the API I'm using, and figured that perhaps I can (a) reduce the number of post-processing required, and (b) reduce the ambiguity e.g. with respect to what the timestamp field is, by directly providing a data frame to the plugin. But I'm not sure what that looks like in practice.

By the way, I have a decent understanding of what a data frame looks like from an abstract perspective (e.g. [from the Grafana docs](https://grafana.com/docs/grafana/latest/developers/plugins/data-frames/)), but not sure what it should look like in the context of the Infinity plugin.

**Comment by svet-b** [comment]
To be a bit more specific, let's say I have a JSON representation of a data frame object, like
```
{
  "name": "Test",
  "fields": [
    {
      "name": "time",
      "type": "time",
      "values": [
        "2023-01-24T15:45:42Z", "2023-01-24T15:45:57Z", "2023-01-24T15:46:13Z", "2023-01-24T15:46:28Z"
      ]
    },
    {
      "name": "field1",
      "type": "number",
      "values": [1, 2, 3, 4]
    },
    {
      "name": "field2",
      "type": "number",
      "values": [5, 6, 7, 8]
    }
  ]
}
```
This is parsed and plotted correctly if I select an inline source and set the format to "As Is":
![Screenshot 2023-01-24 at 6 01 22 PM](https://user-images.githubusercontent.com/12831860/214358620-d657d147-3788-4bd5-9a4e-92c0f3b8dce3.png)
==>
![Screenshot 2023-01-24 at 6 03 30 PM](https://user-images.githubusercontent.com/12831860/214359171-b607fd5f-4384-405d-8220-56b9e965669d.png)


But for example if I set the "Data Frame" format, the data that's shown is quite nonsensical:
![Screenshot 2023-01-24 at 6 04 03 PM](https://user-images.githubusercontent.com/12831860/214359330-7df50d3f-ecc4-4aab-a772-16e5bbe32818.png)
==>
![Screenshot 2023-01-24 at 6 02 15 PM](https://user-images.githubusercontent.com/12831860/214358967-5c4b1a30-348c-4c6c-8421-bcd5e85fd496.png)

I haven't tried with a real API point yet, but wanted to get a handle on a simple use case like this before exploring any further.

**Reply by yesoreyeram** [reply]
Actually you don't need to use the "Format" in most cases and it could have been hidden from the user.. That just exist for various reasons.

* For the timeseries to render properly, data needs some additional post processing like ordering date by time field
* But table/data frame format is almost same and should not have much difference
* For Traces panel to work, data needs to have specific meta data and that's why trace specific formats exits
* As-Is format is just for grafana devs.. who may want to debug different dataframe formats in different panels. 

So the TLDR; is, use timeseries/table only for most common use cases.

**Reply by yesoreyeram** [reply]
to answer you question directly, table and dataframe are same with very negligible difference.

**Reply by svet-b** [reply]
Thanks, this is really helpful context!

**Reply by Lroy** [reply]
I have read through your Q&A on this topic, but I am still confused as to why the data does not show up in the table view.  I used the exact Json that svet-b created in this post.  I'm not quite sure what the issue is.  I need to specify "backend", because I plan to use alerting.

<img width="1625" alt="Screen Shot 2023-02-05 at 12 31 14 PM" src="https://user-images.githubusercontent.com/290178/216835619-de6471da-75ec-407d-9917-fb8da162ae12.png">

**Reply by Lroy** [reply]
BTW, I have modified my /apiv2/status/query to simply return the Json from earlier in this post.

`{
  "name": "Test",
  "fields": [
    {
      "name": "time",
      "type": "time",
      "values": [
        "2023-01-24T15:45:42Z", "2023-01-24T15:45:57Z", "2023-01-24T15:46:13Z", "2023-01-24T15:46:28Z"
      ]
    },
    {
      "name": "field1",
      "type": "number",
      "values": [1, 2, 3, 4]
    },
    {
      "name": "field2",
      "type": "number",
      "values": [5, 6, 7, 8]
    }
  ]
}`

**Comment by Kamilcuk** [comment]
Hello, Is there a simple way to generate this data frame format from python easily? thanks
